pkg/southbound/synchronizer/session.go

```diff
@@ -0,0 +1,247 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package synchronizer
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff"
+
+	"github.com/onosproject/onos-config/pkg/utils"
+
+	devicechange "github.com/onosproject/onos-config/api/types/change/device"
+	"github.com/onosproject/onos-config/pkg/dispatcher"
+	"github.com/onosproject/onos-config/pkg/events"
+	"github.com/onosproject/onos-config/pkg/modelregistry"
+	"github.com/onosproject/onos-config/pkg/southbound"
+
+	devicestore "github.com/onosproject/onos-config/pkg/store/device"
+	"github.com/onosproject/onos-config/pkg/store/mastership"
+
+	devicetype "github.com/onosproject/onos-config/api/types/device"
+	"github.com/onosproject/onos-config/pkg/store/change/device"
+	topodevice "github.com/onosproject/onos-topo/api/device"
+)
+
+const (
+	backoffInterval   = 10 * time.Millisecond
+	maxBackoffTime    = 5 * time.Second
+	mastershipTermKey = "onos-config.mastership.term"
+)
+
+// Session a gNMI session
+type Session struct {
+	mastershipStore           mastership.Store
+	deviceStore               devicestore.Store
+	mastershipState           *mastership.Mastership
+	closeCh                   chan struct{}
+	opStateChan               chan<- events.OperationalStateEvent
+	southboundErrorChan       chan events.DeviceResponse
+	dispatcher                *dispatcher.Dispatcher
+	modelRegistry             *modelregistry.ModelRegistry
+	operationalStateCache     map[topodevice.ID]devicechange.TypedValueMap
+	operationalStateCacheLock *sync.RWMutex
+	deviceChangeStore         device.Store
+	device                    *topodevice.Device
+	target                    southbound.TargetIf
+	cancel                    context.CancelFunc
+	closed                    bool
+	mu                        sync.RWMutex
+}
+
+// open open a new gNMI session
+func (s *Session) open() error {
+	log.Info("Opening a gNMI session")
+	ch := make(chan mastership.Mastership)
+	err := s.mastershipStore.Watch(s.device.ID, ch)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		err := s.updateDeviceState()
+		if err != nil {
+			return
+		}
+	}()
+
+	go func() {
+		connected := false
+		state, _ := s.mastershipStore.GetMastership(s.device.ID)
+		if state != nil {
+			s.mu.Lock()
+			s.mastershipState = state
+			s.mu.Unlock()
+			var currentTerm int
+			if s.device.Attributes == nil {
+				currentTerm = 0
+			} else {
+				currentTerm, _ = strconv.Atoi(s.device.Attributes[mastershipTermKey])
+			}
+
+			if state.Master == s.mastershipStore.NodeID() && uint64(state.Term) >= uint64(currentTerm) {
```

Comment:
@kuujo 
The logic is the same as what we have in onos-ric but with one addition to include mastership term but perhaps you can tell me if there is a better way to address it.

Reply:
It has a bug in it. Fixing it

Reply:
Looks good